backends/vllm/src/block.rs

```diff
@@ -0,0 +1,288 @@
+
+use crate::types::{ReadLock, WriteLock};
+
+/// A mapping between logical and physical KV (Key-Value) blocks for each request.
+///
+/// Each entry in the `BlockTable` represents:
+/// - The corresponding physical block for a logical block
+/// - The number of filled positions in that block
+pub type BlockTable = Vec<SyncPhysicalTokenBlock>;
+
+/// Represents the device on which a block is allocated.
+///
+/// `Cpu`: The block is allocated on the CPU
+///
+/// `Gpu`: The block is allocated on the GPU
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum BlockDevice {
+    Cpu,
+    Gpu,
+}
+
+/// Represents a contiguous chunk of tokens in the logical space.
+///
+/// `LogicalTokenBlock` is used to track the state of corresponding physical blocks
+/// in the KV cache (typically allocated on the GPU). It stores tokens sequentially
+/// from left to right.
+#[derive(Clone, Debug, PartialEq)]
+pub struct LogicalTokenBlock {
+    /// Unique identifier for this block
+    block_number: usize,
+    /// Maximum number of tokens this block can hold
+    block_size: usize,
+    /// Sequence of token IDs, with a maximum length of `block_size`
+    token_ids: Vec<u32>,
+    /// Current number of tokens stored in this block
+    num_tokens: usize,
+    /// Tracing span for observability
+    span: Span,
+}
+
+impl LogicalTokenBlock {
+    /// Constructor
+    pub fn new(block_number: usize, block_size: usize) -> Self {
+        Self {
+            block_number,
+            block_size,
+            token_ids: Vec::with_capacity(block_size),
+            num_tokens: 0,
+            span: info_span!("block"),
+        }
+    }
+
+    /// Getter for `block_number`
+    pub fn block_number(&self) -> usize {
+        self.block_number
+    }
+
+    /// Getter for `block_size`
+    pub fn block_size(&self) -> usize {
+        self.block_size
+    }
+
+    /// Checks if `token_ids` is empty
+    pub fn is_empty(&self) -> bool {
+        self.num_tokens == 0
+    }
+
+    /// Check if `token_ids` is full
+    pub fn is_full(&self) -> bool {
+        self.num_tokens == self.block_size
+    }
+
+    /// Get the number of additional token ids that can be added to the current `LogicalTokenBlock`
+    pub fn get_num_empty_slots(&self) -> usize {
+        self.block_size - self.num_tokens
+    }
+
+    /// Appends a new set of token ids to the current `LogicalTokenBlock`.
+    ///
+    /// # Arguments
+    ///
+    /// * `token_ids` - A slice of u32 values representing the token IDs to be appended.
+    ///
+    /// # Returns
+    ///
+    /// * `Ok(())` if the tokens were successfully appended.
+    /// * `Err(BlockError::AllocationError)` if there isn't enough space in the block.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// let mut block = LogicalTokenBlock::new(0, 10);
+    /// assert!(block.append_tokens(&[1, 2, 3]).is_ok());
+    /// assert!(block.append_tokens(&[4, 5, 6, 7, 8, 9, 10]).is_err());
+    /// ```    #[instrument(skip_all)]
```

Comment:
The instrument should not be in the comment.

Reply:
Great catch up !